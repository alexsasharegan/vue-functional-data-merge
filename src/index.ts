import { VNodeData } from "vue";

/**
 * Intelligently merges data for createElement.
 * Merges arguments left to right, preferring the right argument.
 * Returns new VNodeData object.
 */
function mergeData(...vNodeData: VNodeData[]): VNodeData;
function mergeData(): VNodeData {
  let mergeTarget: VNodeData & { [key: string]: any } = {},
    i: number = arguments.length,
    prop: string,
    event: string;

  // Allow for variadic argument length.
  while (i--) {
    // Iterate through the data properties and execute merge strategies
    // Object.keys eliminates need for hasOwnProperty call
    for (prop of Object.keys(arguments[i])) {
      switch (prop) {
        // Array merge strategy (array concatenation)
        case "class":
        case "style":
        case "directives":
          if (!Array.isArray(mergeTarget[prop])) {
            mergeTarget[prop] = [];
          }
          // Repackaging in an array allows Vue runtime
          // to merge class/style bindings regardless of type.
          mergeTarget[prop] = mergeTarget[prop].concat(arguments[i][prop]);
          break;
        // Space delimited string concatenation strategy
        case "staticClass":
          if (!arguments[i][prop]) {
            break;
          }
          if (mergeTarget[prop] === undefined) {
            mergeTarget[prop] = "";
          }
          if (mergeTarget[prop]) {
            // Not an empty string, so concatenate
            mergeTarget[prop] += " ";
          }
          mergeTarget[prop] += arguments[i][prop].trim();
          break;
        // Object, the properties of which to merge via array merge strategy (array concatenation).
        // Callback merge strategy merges callbacks to the beginning of the array,
        // so that the last defined callback will be invoked first.
        // This is done since to mimic how Object.assign merging
        // uses the last given value to assign.
        case "on":
        case "nativeOn":
          if (!mergeTarget[prop]) {
            mergeTarget[prop] = {};
          }
          for (event of Object.keys(arguments[i][prop] || {})) {
            // Concat function to array of functions if callback present.
            if (mergeTarget[prop][event]) {
              // Insert current iteration data in beginning of merged array.
              mergeTarget[prop][event] = [].concat(
                mergeTarget[prop][event],
                arguments[i][prop][event]
              );
            } else {
              // Straight assign.
              mergeTarget[prop][event] = arguments[i][prop][event];
            }
          }
          break;
        // Object merge strategy
        case "attrs":
        case "props":
        case "domProps":
        case "scopedSlots":
        case "staticStyle":
        case "hook":
        case "transition":
          if (!mergeTarget[prop]) {
            mergeTarget[prop] = {};
          }
          mergeTarget[prop] = { ...arguments[i][prop], ...mergeTarget[prop] };
          break;
        // Reassignment strategy (no merge)
        case "slot":
        case "key":
        case "ref":
        case "tag":
        case "show":
        case "keepAlive":
        default:
          if (!mergeTarget[prop]) {
            mergeTarget[prop] = arguments[i][prop];
          }
      }
    }
  }

  return mergeTarget;
}

export { mergeData };
