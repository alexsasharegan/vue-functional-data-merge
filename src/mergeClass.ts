export type AttrsClass =
  string | string[] | Record<string, boolean> | (string|Record<string, boolean>)[];

function reorganize(attrsClass: AttrsClass): AttrsClass {
  if (!Array.isArray(attrsClass)) {
    return attrsClass;
  }

  const objectIndices: number[] = [];
  for (let i = 0; i < attrsClass.length; i += 1) {
    if (typeof attrsClass[i] === 'object') {
      objectIndices.push(i);
    }
  }
  for (let i = objectIndices.length - 1; i >= 0; i -= 1) {
    const item = attrsClass[objectIndices[i]] as Record<string, boolean>;
    attrsClass.splice(objectIndices[i], 1);
    (attrsClass as Record<string, boolean>[]).push(item);
  }
  return attrsClass;
}

export function mergeClass(
  attrsClass: AttrsClass | undefined,
  classes: AttrsClass,
): AttrsClass {
  if(classes === '') {
    return attrsClass;
  }

  if (attrsClass === undefined) {
    return classes;
  }

  if (typeof attrsClass === 'string') {
    attrsClass = [
      attrsClass,
    ] as string[];
  } else if (!Array.isArray(attrsClass) && typeof attrsClass === 'object') {
    attrsClass = [
      attrsClass,
    ] as (string|Record<string, boolean>)[];
  }

  if (typeof classes === 'string') {
    attrsClass.push(classes);
    return reorganize(attrsClass);
  }

  if (Array.isArray(classes)) {
    for (let i = 0; i < classes.length; i += 1) {
      const item = classes[i];
      if (typeof item !== 'object') {
        const attrItem = attrsClass.find((x) => x === item);
        if (attrItem === undefined) attrsClass.push(item);
      }
    }
  }

  const objectIndex: number = attrsClass.findIndex((x) => typeof x === 'object');
  let classesRecord: Record<string, boolean> | null = null;

  if (!Array.isArray(classes) && typeof classes === 'object') {
    if (objectIndex === -1) {
      (attrsClass as (string|Record<string, boolean>)[]).push(classes);
      return reorganize(attrsClass);
    }

    classesRecord = classes;
  }

  if (classesRecord == null && Array.isArray(classes)) {
    const classObjectIndex: number = classes.findIndex((x) => typeof x === 'object');
    if (classObjectIndex >= 0) classesRecord = classes[classObjectIndex] as Record<string, boolean>;
  }

  if (objectIndex !== -1 && classesRecord != null) {
    Object.keys(classesRecord).forEach((key) => {
      if (Array.isArray(attrsClass) && classesRecord != null) {
        (attrsClass[objectIndex] as Record<string, boolean>)[key] = classesRecord[key];
      }
    });
  } else if (objectIndex === -1 && classesRecord != null) {
    (attrsClass as Record<string, boolean>[]).push(classesRecord);
  }

  return reorganize(attrsClass);
}
