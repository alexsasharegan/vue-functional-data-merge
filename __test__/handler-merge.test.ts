import { VNodeData, VNodeDirective } from "vue"
import { mergeData } from "../src/index"

it("should not mutate original object (issue #2)", () => {
	let def1 = { on: { click() {} } }
	let def2 = { on: { click() {} } }

	let onclick1 = def1.on.click
	let onclick2 = def2.on.click

	let data = mergeData({}, def1, def2)

	expect(def1.on).not.toBe(data.on)
	expect(def2.on).not.toBe(data.on)

	expect(def1.on.click).toBe(onclick1)
	expect(def2.on.click).toBe(onclick2)

	expect(data.on.click).toContain(onclick1)
	expect(data.on.click).toContain(onclick2)
})

it("should set single handlers and concat multi", () => {
	let h1 = console.log
	let h2 = console.info
	let h3 = console.error
	let actual: VNodeData

	actual = mergeData({ class: ["btn", "text-center"] }, { on: { mouseup: h1 } })
	expect(actual).toMatchObject({ on: { mouseup: h1 } })
	expect(Array.isArray(actual.on.mouseup)).toBe(false)

	actual = mergeData({ nativeOn: { mouseup: h1 } }, { class: ["btn", "text-center"] })
	expect(actual).toMatchObject({ nativeOn: { mouseup: h1 } })
	expect(Array.isArray(actual.nativeOn.mouseup)).toBe(false)

	actual = mergeData({ on: { mouseup: h1 } }, { on: { mouseup: h2 } }, { on: { mouseup: h3 } })
	expect(Array.isArray(actual.on.mouseup)).toBe(true)
	expect(actual.on.mouseup).toContain(h1)
	expect(actual.on.mouseup).toContain(h2)
	expect(actual.on.mouseup).toContain(h3)
})

it("should call the right-most argument first", () => {
	let first = 0
	let factory = n => () => {
		if (!first) {
			first = n
		}
	}

	let actual = mergeData(
		{ on: { click: factory(1) } },
		{ on: { click: factory(2) } },
		{ on: { click: factory(3) } },
		{ on: { click: factory(4) } }
	)

	expect(Array.isArray(actual.on.click)).toBe(true)
	for (const fn of actual.on.click) {
		fn()
	}
	expect(first).toBe(4)
	expect(actual).toMatchSnapshot("merge-4")
})
