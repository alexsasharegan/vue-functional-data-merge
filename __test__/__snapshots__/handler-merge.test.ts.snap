// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should call the right-most argument first: merge-4 1`] = `
Object {
  "on": Object {
    "click": Array [
      [Function],
      [Function],
      [Function],
      [Function],
    ],
  },
}
`;
